xarray/core/parallel.py

```diff
@@ -0,0 +1,275 @@
+                graph[task_name] = (tuple, [variable.dims, chunk])
+            else:
+                # numpy array with possibly chunked dimensions
+                # index into variable appropriately
+                subsetter = dict()
+                for dim in variable.dims:
+                    if dim in chunk_index_dict:
+                        which_chunk = chunk_index_dict[dim]
+                        subsetter[dim] = slice(
+                            chunk_index_bounds[dim][which_chunk],
+                            chunk_index_bounds[dim][which_chunk + 1],
+                        )
+
+                subset = variable.isel(subsetter)
+                task_name = (name + dask.base.tokenize(subset),) + v
+                graph[task_name] = (tuple, [subset.dims, subset])
+
+            # this task creates dict mapping variable name to above tuple
+            if name in dataset.data_vars:
+                data_vars.append([name, task_name])
+            if name in dataset.coords:
+                coords.append([name, task_name])
+
+        from_wrapper = (gname,) + v
+        graph[from_wrapper] = (
+            _wrapper,
+            func,
+            (Dataset, (dict, data_vars), (dict, coords), dataset.attrs),
+            input_is_array,
+            args,
+            kwargs,
+        )
+
+        # mapping from variable name to dask graph key
+        var_key_map = {}
+        for name, variable in template.variables.items():
+            var_dims = variable.dims
+            # cannot tokenize "name" because the hash of <this-array> is not invariant!
+            # This happens when the user function does not set a name on the returned DataArray
+            gname_l = "%s-%s" % (gname, name)
+            var_key_map[name] = gname_l
+
+            key = (gname_l,)
+            for dim in var_dims:
+                if dim in chunk_index_dict:
+                    key += (chunk_index_dict[dim],)
+                else:
+                    # unchunked dimensions in the input have one chunk in the result
+                    key += (0,)
+
+            graph[key] = (operator.getitem, from_wrapper, name)
+
+    graph = HighLevelGraph.from_collections(name, graph, dependencies=[dataset])
+
+    result = Dataset()
+    for var, key in var_key_map.items():
+        # indexes need to be known
+        # otherwise compute is called when DataArray is created
+        if var in indexes:
+            result[var] = indexes[var]
+            continue
+
+        dims = template[var].dims
+        var_chunks = []
+        for dim in dims:
+            if dim in input_chunks:
+                var_chunks.append(input_chunks[dim])
+            else:
+                if dim in indexes:
+                    var_chunks.append((len(indexes[dim]),))
+
+        data = dask.array.Array(
+            graph, name=key, chunks=var_chunks, dtype=template[var].dtype
+        )
+        result[var] = DataArray(data=data, dims=dims, name=var)
```

Comment:
``result[var] = dims, data``, and again you're accidentally moving non-index coords to data_vars